eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/DualFundingHandlers.scala

```diff
@@ -59,25 +59,39 @@ trait DualFundingHandlers extends CommonFundingHandlers {
     }
   }
 
+  def pruneCommitments(w: WatchFundingConfirmedTriggered, d: DATA_WAIT_FOR_DUAL_FUNDING_CONFIRMED): Option[DualFundingTx] = {
+    val allFundingTxs: Seq[DualFundingTx] = DualFundingTx(d.fundingTx, d.commitments) +: d.previousFundingTxs
+    // We can forget other funding attempts now that one of the funding txs is confirmed.
+    val otherFundingTxs = allFundingTxs.filter(_.commitments.fundingTxId != w.tx.txid).map(_.fundingTx)
+    rollbackDualFundingTxs(otherFundingTxs)
+    // We find which funding transaction got confirmed.
+    allFundingTxs.find(_.commitments.fundingTxId == w.tx.txid) match {
+      case Some(dft: DualFundingTx) =>
+        watchFundingTx(dft.commitments)
+        context.system.eventStream.publish(TransactionConfirmed(d.channelId, remoteNodeId, w.tx))
+        Some(dft)
+      case None =>
+        // An unknown funding tx has been confirmed, this should never happen. Note that this isn't a case of
+        // ERR_INFORMATION_LEAK, because here we receive a response from the watcher from a WatchConfirmed that we put
+        // ourselves and somehow forgot.
+        log.error(s"internal error: the funding tx that confirmed doesn't match any of our funding txs: ${w.tx.bin}")
+        None
+    }
+  }
+
+  /**
+   * We could just ignore the event and handle it after the next reconnection, but this allows us to eagerly rollback
+   * and forget deprecated funding transactions.
```

Comment:
This comment should mention that it's more than that, it's a matter of funds safety: if we didn't handle this event while offline, our peer could force-close and never reconnect, and we wouldn't claim our funds back.

Reply:
Fixed in [07e6465](https://github.com/ACINQ/eclair/pull/2557/commits/07e646591988dd28c52070ac65589e3ce6c86cc1).

Interestingly, in parallel I also added an additional factorization in commit [5c0c189](https://github.com/ACINQ/eclair/pull/2557/commits/5c0c189e9c92dc04c83a3a06ce57b430a1d7a25a) which made the `watchFundingTx` appear expliclty in `handleDualFundingConfirmedOffline`. I wasn't too sure about that commit, but this makes me think it's directionnally correct, because the critical operation is more obvious, while it was hidden before.